### PR TITLE
More work on counts for paper routes as part of #125

### PR DESCRIPTION
**Pull request by josmas**
- a bit of refactoring on papers/badges[/count]
- new endpoints for three last routes in papers, with tests.

This is part of the work in #125 for the /papers route.

### CODE REVIEW

**Comment by josmas** [comment]
I have also added counts for the _/badges_ route.

**Comment by josmas** [comment]
Last piece of work for #125 is in now (user count route). I have also refactored a bit so that the code is consistent with other routes in both _/users_ and _/papers_.

@acabunoc this is a large change, take your time reviewing; there are tests for all the routes, which should make things easier.

**Comment by acabunoc** [comment]
Thanks @josmas! I'll try to review today :)

File: templates/pages/home.jsx [hunk 1]
```diff
@@ -28,12 +28,23 @@ var Home = React.createClass({
               <li>Get all badges we issue</li>
             </ul>
           </li>
+          <li>GET <a href="/badges/count">/badges/count</a>
```
Comment (acabunoc) [inline templates/pages/home.jsx]: Thanks for adding documentation both here *and* the README. I know it's a pain!

File: src/routes/papers/index.js [hunk 2]
```diff
@@ -71,10 +93,24 @@ function getUserBadgesByType(request, response) {
...
     '_2': request.params.doi2
-  }), request, response);
+  });
+}
+
+function getUserBadgesByBadge(request, response) {
+  returnBadges(getUserBadgesByType(request, response), request, response);
+}
+
+function getUserBadgesByBadgeCount(request, response) {
```
Comment (acabunoc) [inline src/routes/papers/index.js]: This should never be > 1 (you can only claim each badge type once per paper)

Comment (josmas) [inline src/routes/papers/index.js]: cool; I can change the asserts in the tests. Can you have a look at those numbers? I didn't know exactly what the expected values would be.

Comment (acabunoc) [inline src/routes/papers/index.js]: most of the numbers would change over time as more badges are issues! Except for this route, which should always be 0 or 1.

File: src/routes/papers/index.js [hunk 3]
```diff
@@ -41,28 +34,57 @@ function getBadgeCount(request, response) {
...
-  returnBadges(badgerService.getBadges(null, request.params.badge, {
+  return badgerService.getBadges(null, request.params.badge, {
     '_1': request.params.doi1,
     '_2': request.params.doi2
-  }), request, response);
+  });
 }
 
-function getUserBadges(request, response) {
+function getBadgesByBadge(request, response) {
```
Comment (acabunoc) [inline src/routes/papers/index.js]: I like how the function names are consistent now! I don't like how the name is more confusing now.

I *think* I'm okay with this change, but I'll think a bit more.

File: test/integration/app.js [hunk 4]
```diff
@@ -59,7 +77,7 @@ describe('Integration test against the real Badge server', function () {
         .expect(200, done);
   });
 
-  it('get the number of badges earned by a user', function (done) {
+  it('get a count of all badge instances earned by a user', function (done) {
```
Comment (acabunoc) [inline test/integration/app.js]: +1 consistency! 👍 💯

File: src/routes/papers/index.js (hunk 2 above)
Comment (josmas) [inline src/routes/papers/index.js]: I see. Good, so I'll change the assert to `<= 1` for this one. won't move things around till you finish the review though, so we don't get lost in changes.

**Comment by acabunoc** [comment]
This looks good to me! R+ once you update the assert for `/papers/:doi1/:doi2/users/:orcid/badges/:badge/count`

Feel free to merge once that's in :) Thanks for this, it's a huge lift!

**Pull request merged**